src/Core/Checkout/Cart/Order/OrderConverter.php

```diff
@@ -391,10 +404,21 @@ public function assembleSalesChannelContext(OrderEntity $order, Context $context
         return $salesChannelContext;
     }
 
-    private function convertDeliveries(OrderDeliveryCollection $orderDeliveries, LineItemCollection $lineItems): DeliveryCollection
+    private function convertDeliveries(?string $primaryOrderDeliveryId, OrderDeliveryCollection $orderDeliveries, LineItemCollection $lineItems): DeliveryCollection
     {
+        // Ensure primary delivery is first, so `$deliveries->first()` returns the primary delivery.
+        $keys = array_unique([$primaryOrderDeliveryId, ...$orderDeliveries->getKeys()]);
+
+        if ($primaryOrderDeliveryId === null) {
+            $keys = \array_slice($keys, 1);
+        }
+
         $cartDeliveries = new DeliveryCollection();
-        foreach ($orderDeliveries as $orderDelivery) {
+        foreach ($keys as $id) {
+            if (!$id || !$orderDelivery = $orderDeliveries->get($id)) {
+                continue;
+            }
```

Comment:
Every ID should exist, if not e.g. at least tests should fail
```suggestion
            if (!$orderDelivery = $orderDeliveries->get($id)) {
                throw Exception
            }
```